README.md
# Empire of Thorns or EOT

Hi! This game was modified mainly by myself. The idea was to get the original open source game and merge it with the two other versions, main one called original (Stendhal Game), the second one is a chinese version (YRPG), and the third is the polish one (Polska Online).

While the main version is in the release 1.33, the chinese version was in 0.80 and the polish one was in 0.60, so the biggest challenge was to do the updates and merge the main funtionabilities to the other versions.

This version isn't done yet, still needs polish, new features, translation of the polish version, and update from java 8 to java 9 wich has deprecated the Observer class, and that the version 1.26 of the main version uses a lot.

Without further ado, I will leave with the original server description of the game. 
 
## Stendhal

Stendhal is a fun friendly and free multiplayer online adventure game. Start playing, get hooked... Get the source code, and add your own ideas...

Stendhal has a huge and rich world. You can explore cities, forest, mountains, mines, plains, caves and dungeons. You could meet over 250 NPCs, many will give you tasks and quests for valuable experience, shiny gold, but mainly a warm fuzzy feeling. If you just want to fight that helps too - the cities could get overrun by monsters without you!

Your character will develop and grow and with each new level up become stronger and better. With the money you acquire you can buy new items and improve your armour and weapons, trade with friends, or look for what other players are giving away in the 'community chest'.

Stendhal runs on Windows, Mac and Linux, written using Java 1.5 and the Java2D environment. The main game server has a website at  [http://stendhalgame.org](http://stendhalgame.org/).

## Playing

If you want to play the Stendhal, you need the Stendhal client, which is the stendhal-1.xx.zip file.

Then unzip the client zip into a new folder.

You can double click on stendhal-starter.jar and the client will run.

If you prefer to run from command line, then just execute this command in the stendhal folder:

-   java -jar stendhal-starter.jar

There is a manual at  [http://stendhalgame.org/wiki/Stendhal_Manual](http://stendhalgame.org/wiki/Stendhal_Manual), and there is Help from the client main menu too.














doc/CHANGES.txt
Changelog
---------
---------

1.33

*webclient*
- long touching to simulate right mouse button
- added support for additional entities: rented signs, blackboards, blocks, house doors, doors
- added online/offline icons on buddy list
- added commands /drop, /adminlevel


*world*
- new regions/maps
  - Atlantis
  - Deniran Tannery
- new NPCs
  - Jimbo (Deniran Grocery Store)
  - Skinner Rawhide (Denrian Tannery)
  - Zelan (Atlantis)
- NPC changes
  - elvish hat added to items that Lupos buys
  - Adena sells olive oil & vinegar
- new items:
  - pelt (ingredient to make money pouch)
  - leather thread (ingredient to make money pouch)
  - leather needle (ingredient to make money pouch)
  - imperial ring (+1 ATK)
  - placid ring (shock resist: 0.33)
  - magic plate legs
  - magic plate boots
  - rotary cutter (tool for making leather thread)
  - royal armor, royal boots, royal cloak, royal helmet, royal legs, royal shield, royal dagger
  - unicorn horn
- new enemies:
  - bull elephant, musth elephant
  - giant bat, giant killer bat
  - poison shroom
  - polar bear
  - tiny scarab, water scarab
  - water demon
  - white unicorn, black unicorn, golden unicorn
  - baby pegasus, bronze pegasus, ivory pegasus, ebony pegasus
  - solar centaur, glacier centaur, centaur king
  - woolly mammoth
  - giant gnome, giant gnomess
  - neo kraken
  - rysin dragon
- enemy changes
  - dimensions
    - changed ent, old ent, & entwife to 2x1 to match sprites better
  - new item drops
    - imperial ring
      - imperial veteran (@ 0.017)
      - imperial elite archer (@ 0.05)
    - insulated ring (cobra: @ 0.08)
    - placid ring
      - gargoyle (@ 0.005)
      - ghost (@ 0.02)
      - incorporeal armor (@ 0.3)
      - musth elephant (@ 0.5)
      - polar bear (@ 0.1)
      - rysin dragon (@ 2.5)
      - water scarab (@ 0.75)
      - werewolf (@ 0.5)
    - magic plate boots (cobra: @ 0.01)
    - magic plate legs (king cobra: @ 0.25)
    - pelt
      - deer (@ 1.0)
      - beaver (@ 1.0)
      - bear (@ 2.0)
      - black bear (@ 4.0)
    - poison (cobra: 2-5 @ 20.0)
  - sounds
    - replaced kobold moan sounds with bark
- players can enable a new pouch slot to carry money
- quest changes
  - items for Kirdneh Museum quest categorized for player level
    - "easy": any level
    - "medium": for players level 51+
    - "hard": for players level 151+
  - items added to Daily Item quest
    - leather needle
    - leather thread (20)
    - pelt (5)
  - items added to Kirdneh Museum quest
    - medicinal ring (medium)
    - scroll eraser (medium)
    - ugmash (easy)
- new quests
  - Unicorn Horns for Zelan
- new achievements:
  - Westerner: visit all outside zones in Deniran region (easy score)
  - Spelunker: visit all underground zones in Deniran region (hard score)
  - Cheese Wiz: loot 2,000 cheese (easy score)
  - Ham Hocks: loot 2,500 ham (easy score)
  - Goldshower: loot 10,000 money (easy score)
  - Green Thumb: harvest 20 of each type of growable flower (easy score)
  - David vs. Goliath: kill 20 of each type of giant solo (medium score)
  - Heavenly Wrath: kill 100 of each type of angel (hard score)
  - Silver Bullet: kill 500 werewolves (medium score)
  - Herbal Practitioner: harvest 20 of each type of herb found growing in Faiumoni (easy score)
  - It's Happy Hour Somewhere: purchase 100 bottles of beer & 100 glasses of wine (easy score)
  - Serenade the Siren: kill 10,000 gem mermaids (hard score)
  - Imperialist Lackey: finish Kill Blordroughs quest 5 times (medium score)
  - Imperialist Soldier: finish Kill Blordoughs quest 25 times (hard score)
- new music
  - -1_semos_yeti_cave
  - levels -2 & -3 of Semos Catacombs
  - Kikareukin Caves
  - Deniran City
  - Deniran Forest
  - Deniran Caves
  - Atlantis
  - Semos Village, Plains, & Road
  - Nalwor River
  - taverns & inns in Ados, Fado, & Kirdneh cities
- status effects
  - shock resistance also alters duration of status

*grapics*
- made dark dagger image more distinct from night dagger
- changed beer image to fit description better
- restored old banana, grapes, pear, & watermelon images
- new item images:
  - golden helmet
  - viking helmet
  - aventail
  - scroll eraser
  - studded boots
  - trophy helmet


*classic client*
- entities with "active_idle" attribute will cycle animation even if stopped
- minimap recognizes Door entity as Portal

*scripting*
- added Lua support

*fixes*
- support for stairs portals in Tiled with continuous movement
- fixed Door & LockedDoor entities not handling continuous movement
- fixed cannot complete Traps for Klaas quest when Antivenom Ring quest activated
- fixed cannot enter apothecary lab if Traps for Klaas quest restarted
- enemy/creature shadow can be set with "shadow" value under "type" tag
- removed multiple messages about no longer being shocked
- fixed veil not being removed if renting from Timothy while wearing wedding gown
- slash action "/removedetail" can be executed to manually remove the detail layer from a character
- mithril cloak quest can be reset to an earlier state in case of lost items



1.32
*world*
- new enemies
  - amethyst mermaid, emerald mermaid, ruby mermaid, & sapphire mermaid
  - cannibal warrior
  - legendary ice elemental
  - turtle
- new NPCs
  - Jameson (Apothecary Lab)
  - Wanda (Denrian Potions Shop)

src/games/stendhal/server/entity/Entity.java
/* $Id$ */
/***************************************************************************
 *                      (C) Copyright 2003 - Marauroa                      *
 ***************************************************************************
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
package games.stendhal.server.entity;

import static games.stendhal.server.core.engine.Translate.getText;

import java.awt.Point;
import java.awt.Rectangle;
import java.awt.geom.Rectangle2D;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

import org.apache.log4j.Logger;

import com.google.common.base.Predicate;
import com.google.common.collect.Iterators;

import games.stendhal.common.ItemTools;
import games.stendhal.common.constants.Events;
import games.stendhal.common.grammar.Grammar;
import games.stendhal.server.core.engine.StendhalRPZone;
import games.stendhal.server.core.pathfinder.Node;
import games.stendhal.server.entity.slot.EntitySlot;
import games.stendhal.server.entity.slot.SlotNameInList;
import games.stendhal.server.entity.slot.Slots;
import marauroa.common.game.Definition;
import marauroa.common.game.Definition.Type;
import marauroa.common.game.RPClass;
import marauroa.common.game.RPObject;
import marauroa.common.game.RPSlot;

public abstract class Entity extends RPObject implements Killer {
	/**
	 * The logger.
	 */
	private static final Logger logger = Logger.getLogger(Entity.class);


	protected Rectangle2D.Double area = new Rectangle2D.Double();

	private int x;

	private int y;

    // Initial coordinates
    private Point origin;

	/**
	 * Amount of resistance this has with other entities (0-100).
	 */
	private int resistance;

	private StendhalRPZone zone;
	private StendhalRPZone lastZone;

	public Entity(final RPObject object) {
		super(object);

		if (!has("x")) {
			put("x", 0);
		}

		if (!has("y")) {
			put("y", 0);
		}

		if (!has("width")) {
			put("width", 1);
		}

		if (!has("height")) {
			put("height", 1);
		}

		if (!has("resistance")) {
			put("resistance", 100);
		}

		if (!has("visibility")) {
			put("visibility", 100);
		}

		update();
	}

	public Entity() {
		put("x", 0);
		put("y", 0);

		x = 0;
		y = 0;

		setSize(1, 1);
		area.setRect(x, y, 1, 1);

		setResistance(100);
		setVisibility(100);
	}

	public static void generateRPClass() {
		final RPClass entity = new RPClass("entity");

		// Some things may have a textual description
		entity.addAttribute("description", Type.LONG_STRING, Definition.HIDDEN);

		// TODO: Try to remove this attribute later (at DB reset?)
		entity.addAttribute("type", Type.STRING);

		/**
		 * Resistance to other entities (0-100). 0=Phantom, 100=Obstacle.
		 */
		entity.addAttribute("resistance", Type.BYTE, Definition.VOLATILE);

		entity.addAttribute("x", Type.SHORT);
		entity.addAttribute("y", Type.SHORT);

		/*
		 * The size of the entity (in world units).
		 */
		entity.addAttribute("width", Type.SHORT, Definition.VOLATILE);
		entity.addAttribute("height", Type.SHORT, Definition.VOLATILE);

		/*
		 * Obsolete and ignored by the client. Do not use.
		 */
		entity.addAttribute("server-only", Type.FLAG, Definition.VOLATILE);

		/*
		 * The current overlayed client effect.
		 */
		entity.addAttribute("effect", Type.STRING, Definition.VOLATILE);

		/*
		 * The visibility of the entity drawn on client (0-100). 0=Invisible,
		 * 100=Solid. Useful when mixed with effect.
		 */
		entity.addAttribute("visibility", Type.INT, Definition.VOLATILE);

		// cursor
		entity.addAttribute("cursor", Type.STRING);

		// menu (Make a wish,use)
		entity.addAttribute("menu", Type.STRING, Definition.VOLATILE);

		// sound events
		entity.addRPEvent(Events.SOUND, Definition.VOLATILE);
		// graphical effects
		entity.addRPEvent(Events.IMAGE, Definition.VOLATILE);
		entity.addRPEvent(Events.PUBLIC_TEXT, Definition.VOLATILE);
	}


	public void update() {
		final int oldX = x;
		final int oldY = y;
		boolean moved = false;

		if (has("x")) {
			x = getInt("x");
			area.x = x;

			if (x != oldX) {
				moved = true;
			}
		}

		if (has("y")) {
			y = getInt("y");
			area.y = y;

			if (y != oldY) {
				moved = true;
			}
		}

		if (moved && (zone != null)) {
			onMoved(oldX, oldY, x, y);
		}

		if (has("height")) {
			area.height = getInt("height");
		}

		if (has("width")) {
			area.width = getInt("width");
		}

		if (has("resistance")) {
			resistance = getInt("resistance");
		}

	}

	public boolean hasDescription() {
		if (has("description")) {
			return ((getDescription() != null) && (getDescription().length() > 0));
		}
		return (false);
	}

	public void setDescription(final String text) {
		if (text == null) {
			put("description", "");
		} else {
			put("description", text);
		}

	}

	public String getDescription() {
		String description = "";
		if (has("description")) {
			description = get("description");
		}
		return getText(description);
	}

	/**
	 * Get the nicely formatted entity title/name.
	 *
	 * @return The title, or <code>null</code> if unknown.
	 */
	public String getTitle() {
		if (has("subclass")) {
			return ItemTools.itemNameToDisplayName(get("subclass"));
		} else if (has("class")) {
			return ItemTools.itemNameToDisplayName(get("class"));
		} else if (has("type")) {
			return ItemTools.itemNameToDisplayName(get("type"));
		} else {
			return null;
		}
	}

	/**
	 * Get the entity X coordinate.
	 *
	 * @return The X coordinate.
	 */
	public int getX() {
		return x;
	}

	/**
	 * Get the entity Y coordinate.
	 *
	 * @return The Y coordinate.
	 */
	public int getY() {
		return y;
	}

	/**
	 * Get the zone this entity is in.
	 *
	 * @return A zone, or <code>null</code> if not in one.
	 */
	public StendhalRPZone getZone() {
		// Use onAdded()/onRemoved() to grab a reference to the zone and save
		// as a attribute.
		// During zone transfer zone is set to null for a short period of time
		// which causes lots of problems, so we use the old zone until the new
		// one is set.
		return lastZone;
	}

	/**
	 * Is this entity not moving?
	 *
	 * @return true, if it stopped, false if it is moving
	 */
	public boolean stopped() {
		return true;
	}

	/**
	 * Get the resistance this has on other entities (0-100).
	 *
	 * @return The amount of resistance, or 0 if in ghostmode.
	 */
	public int getResistance() {
		return resistance;
	}

	/**
	 * Get the resistance between this and another entity (0-100).
	 * @param entity other entity to be evaluated
	 *
	 * @return The amount of combined resistance.
	 */
	public int getResistance(final Entity entity) {
		return ((getResistance() * entity.getResistance()) / 100);
	}


	/**
	 * Determine if this is an obstacle for another entity.
	 *
	 * @param entity
	 *            The entity to check against.
	 *
	 * @return <code>true</code> if very high resistance.
	 */
	public boolean isObstacle(final Entity entity) {
		// > 95% combined resistance = obstacle
		return (getResistance(entity) > 95);
	}

	/**
	 * This returns square of the distance between this entity and the given
	 * one. We're calculating the square because the square root operation would
	 * be expensive. As long as we only need to compare distances, it doesn't
	 * matter if we compare the distances or the squares of the distances (the
	 * square operation is strictly monotonous for positive numbers).
	 *
	 * @param other
	 *            the entity to which the distance should be calculated
	 * @return double representing the squared distance
	 */
	public final double squaredDistance(final Entity other) {
		final Rectangle2D otherArea = other.getArea();
		final double otherMiddleX = otherArea.getCenterX();
		final double otherMiddleY = otherArea.getCenterY();

		final Rectangle2D thisArea = getArea();
		final double thisMiddleX = thisArea.getCenterX();
		final double thisMiddleY = thisArea.getCenterY();

		double xDistance = Math.abs(otherMiddleX - thisMiddleX) - (area.getWidth() + other.area.getWidth()) / 2;
		double yDistance = Math.abs(otherMiddleY - thisMiddleY) - (area.getHeight() + other.area.getHeight()) / 2;

		if (xDistance < 0) {
			xDistance = 0;
		}
		if (yDistance < 0) {
			yDistance = 0;
		}
		return xDistance * xDistance + yDistance * yDistance;
	}

	/**
	 * This returns square of the distance from this entity to a specific point.
	 * We're calculating the square because the square root operation would be
	 * expensive. As long as we only need to compare distances, it doesn't
	 * matter if we compare the distances or the squares of the distances (the
	 * square operation is strictly monotonous for positive numbers).
	 *
	 * @param x
	 *            The horizontal coordinate of the point
	 * @param y
	 *            The vertical coordinate of the point
	 * @return double representing the squared distance
	 */
	public final double squaredDistance(final int x, final int y) {


		final double otherMiddleX = x + 0.5;
		final double otherMiddleY = y + 0.5;


		final Rectangle2D thisArea = getArea();

		final double thisMiddleX = thisArea.getCenterX();
		final double thisMiddleY = thisArea.getCenterY();


		double xDistance = Math.abs(otherMiddleX - thisMiddleX) - (area.getWidth() + 1) / 2;
		double yDistance = Math.abs(otherMiddleY - thisMiddleY) - (area.getHeight() + 1) / 2;

		if (xDistance < 0) {
			xDistance = 0;
		}
		if (yDistance < 0) {
			yDistance = 0;
		}
		return xDistance * xDistance + yDistance * yDistance;
	}

	/**
	 * Checks whether the given entity is directly next to this entity. This
	 * method may be optimized over using nextTo(entity, 0.25).
	 *
	 * @param entity
	 *            The entity
	 *
	 * @return <code>true</code> if the entity is next to this.
	 */
	public boolean nextTo(final Entity entity) {
		// For now call old code (just a convenience function)
		return nextTo(entity, 0.25);
	}

	/**
	 * Checks whether the given entity is near this entity.
	 *
	 * @param entity
	 *            the entity
	 * @param step
	 *            The maximum distance
	 * @return true iff the entity is at most <i>step</i> steps away
	 */
	public boolean nextTo(final Entity entity, final double step) {
		// To check the overlapping between 'this' and the other 'entity'
		// we create two temporary rectangle objects and initialise them
		// with the position of the two entities.
		// The size is calculated from the original objects with the additional
		// 'step' distance on both sides of the two rectangles.
		// As the absolute position is not important, 'step' need not be
		// subtracted from the values of getX() and getY().
		final Rectangle2D thisArea = new Rectangle2D.Double(x - step, y - step, area.getWidth()
				+ 2 * step, area.getHeight() + 2 * step);

		return thisArea.intersects(entity.getArea());
	}

	/**
	 * Get horizontal & vertical nodes immediately adjacent to entity's
	 * current X,Y coordinates.
	 *
	 * NOTE: This does not compensate for the entity's width & height,
	 *       it is only meant to get the positions that the entity
	 *       could potentially move to.
	 */
	public List<Node> getAdjacentNodes() {
		List<Node> nodes = new ArrayList<>();
		final int x = getX();
		final int y = getY();

		for (final int lat: Arrays.asList(x-1, x+1)) {
			nodes.add(new Node(lat, y));
		}
		for (final int lon: Arrays.asList(y-1, y+1)) {
			nodes.add(new Node(x, lon));
		}

		return nodes;
	}

	/**
	 * Get the area this object currently occupies.
	 *
	 * @return A rectangular area.
	 */
	public Rectangle2D getArea() {
		return area;
	}

	/**
	 * Returns the area used by this entity.
	 *
	 * @param ex
	 *            x
	 * @param ey
	 *            y
	 * @return rectangle for the used area
	 */
	public Rectangle2D getArea(final double ex, final double ey) {
		final Rectangle2D tempRect = new Rectangle.Double();
		tempRect.setRect(ex, ey, area.getWidth(), area.getHeight());
		return tempRect;
	}


	/**
	 * Called when this object is added to a zone.
	 *
	 * @param zone
	 *            The zone this was added to.
	 */
	public void onAdded(final StendhalRPZone zone) {
		if (this.zone != null) {
			logger.error("Entity added while in another " + zone + ": " + this, new Throwable());
		}

		this.zone = zone;
		this.lastZone = zone;
	}

	/**
	 * Notification of intra-zone position change.
	 *
	 * @param oldX
	 *            The old X coordinate.
	 * @param oldY
	 *            The old Y coordinate.
	 * @param newX
	 *            The new X coordinate.
	 * @param newY
	 *            The new Y coordinate.
	 */
	protected void onMoved(final int oldX, final int oldY, final int newX, final int newY) {
		// sub classes can implement this method
	}

	/**
	 * Called when this object is being removed from a zone.
	 *
	 * @param zone
	 *            The zone this will be removed from.
	 */
	public void onRemoved(final StendhalRPZone zone) {
		if (this.zone != zone) {
			logger.error("Entity removed from wrong zone " + zone + " but it thinks it is in " + this.zone + ": ", new Throwable());
		}

		this.zone = null;
	}

	/**
	 * Notifies the StendhalRPWorld that this entity's attributes have changed.
	 *
	 */
	public void notifyWorldAboutChanges() {
		/*
		 * Only possible if in a zone. This does *NOT* use getZone(), because
		 * that can return a zone that currently does not really contain this
		 * entity, and MarauroaRPZone.modify() assumes the object modified is
		 * really contained there. Modifying an entity that's not really in the
		 * zone results in a perception for an entity that the client does not
		 * recognize.
		 */
		if (zone != null) {
			zone.modify(this);
		}
	}

	/**
	 * Describes the entity (if a players looks at it).
	 *
	 * @return description from the players point of view
	 */
	public String describe() {
		if (hasDescription()) {
			return getDescription();
		}

		return "You see " + getDescriptionName(false) + ".";
	}

	/**
	 * Returns the name or something that can be used to identify the entity for
	 * the player.
	 *
	 * @param definite
	 *            true for "the" and false for "a/an" in case the entity has no
	 *            name
	 * @return name
	 *
	 */
	public String getDescriptionName(final boolean definite) {
		if (has("subclass")) {
			return Grammar.article_noun(ItemTools.itemNameToDisplayName(get("subclass")), definite);
		} else if (has("class")) {
			return Grammar.article_noun(ItemTools.itemNameToDisplayName(get("class")), definite);
		} else {
			String ret = "something indescribably strange";
			if (has("type")) {
				ret += " of type " + ItemTools.itemNameToDisplayName(get("type"));
			}
			if (has("id")) {
				ret += " with id " + get("id");
			}
			if (has("zone")) {
				ret += " in zone " + get("zone");
			}
			logger.error("Missing description for " + this);
			return ret;
		}
	}

	/**
	 * Get the entity height.
	 *
	 * @return The height (in world units).
	 */
	public double getHeight() {
		return area.getHeight();
	}

	/**
	 * Get the entity width.
	 *
	 * @return The width (in world units).
	 */
	public double getWidth() {
		return area.getWidth();
	}

	/**
	 * Set the entity class.
	 *
	 * @param clazz
	 *            The class name.
	 */
	public final void setEntityClass(final String clazz) {
		put("class", clazz);
	}

	/**
	 * Set the entity sub-class.
	 *
	 * @param subclazz
	 *            The sub-class name.
	 */
	public final void setEntitySubclass(final String subclazz) {
		put("subclass", subclazz);
	}

	/**
	 * Sets the entity position.
	 *
	 *
	 * <p>
	 * This calls <code>onMoved()</code>. <strong>Note: When placing during a
	 * zone change, this call should be done after being removed from the old
	 * zone, but before adding to the zone to prevent an erroneous position jump
	 * in the zone.</strong>
	 *
	 * @param x
	 *            The x position (in world units).
	 * @param y
	 *            The y position (in world units).
	 */
	public final void setPosition(final int x, final int y) {
		final int oldX = this.x;
		final int oldY = this.y;
		boolean moved = false;

        // Set the original position of the entity
        if (origin == null) {
            origin = new Point(x, y);
        }

		if (x != oldX) {
			this.x = x;
			area.x = x;
			put("x", x);
			moved = true;
		}

		if (y != oldY) {
			this.y = y;
			area.y = y;
			put("y", y);
			moved = true;
		}

		if (moved && (zone != null)) {
			onMoved(oldX, oldY, x, y);
		}
	}

	/**
	 *
	 * @return The initial position of the entity
	 */
	public final Point getOrigin() {
	    return origin;
	}

	/**
	 * Set resistance this has with other entities.
	 *
	 * @param resistance
	 *            The amount of resistance (0-100).
	 */
	public final void setResistance(final int resistance) {
		this.resistance = resistance;
		put("resistance", resistance);
	}

	/**
	 * Set the entity size.
	 *
	 * @param width
	 *            The width (in world units).
	 * @param height
	 *            The height (in world units).
	 */
	public void setSize(final int width, final int height) {
		this.area.width = width;
		put("width", width);

		this.area.height = height;
		put("height", height);
	}

	/**
	 * gets the name of the mouse cursor or <code>null</code>.
	 *
	 * @return name of the mouse cursor or <code>null</code>.
	 */
	public String getCursor() {
		if (has("cursor")) {
			return get("cursor");
		}
		return null;
	}

	/**
	 * sets the name of the mouse cursor
	 *
	 * @param cursor name of cursor
	 */
	public void setCursor(String cursor) {
		if (cursor == null) {
			remove("cursor");
		} else {
			put("cursor", cursor);
		}
	}

	/**
	 * Set the entity's visibility.
	 *
	 * @param visibility
	 *            The visibility (0-100).
	 */
	public final void setVisibility(final int visibility) {
		put("visibility", visibility);
	}

	/**
	 * Check if the other Entity is near enough to be in sight on the client screen.
	 *
	 * @param other
	 * @return true if near enough
	 */
	public boolean isInSight(final Entity other) {
		if (other != null) {
			if (other.getZone() == getZone()) {
				// check distance: 640x480 client screen size for 32x32 pixel tiles
				// -> makes 20x15 tiles screen size
				if ((Math.abs(other.getX() - x) <= 20)
						&& (Math.abs(other.getY() - y) <= 15)) {
					return true;
				}
			}
		}

		return false;
	}

	/**
	 * gets the named entity slot
	 *
	 * @param name name of entity slot
	 * @return EntitySlot or <code>null</code>
	 */
	public EntitySlot getEntitySlot(String name) {
		RPSlot slot = super.getSlot(name);
		if (!(slot instanceof EntitySlot)) {
			return null;
		}
		return (EntitySlot) slot;
	}

	/**
	 * an iterator over slots
	 *
	 * @param slotTypes slot types to include in the iteration
	 * @return Iterator
	 */
	public Iterator<RPSlot> slotIterator(Slots slotTypes) {
		Predicate<RPSlot> p = new SlotNameInList(slotTypes.getNames());
		return Iterators.filter(slotsIterator(), p);
	}

	/**
	 * an Iterable over slots
	 *
	 * @param slotTypes slot types to include in the iteration
	 * @return Iterable
	 */
	public Iterable<RPSlot> slots(final Slots slotTypes) {
		return new Iterable<RPSlot>() {
			@Override
			public Iterator<RPSlot> iterator() {
				return slotIterator(slotTypes);
			}
		};
	}

	/**
	 * The menu to display on the client in the format:
	 * <pre>
	 *    Display Name 1|action1,
	 *    Display Name 2|action2
	 * </pre>
	 *
	 * @param menu menu string
	 */
	public void setMenu(String menu) {
		put("menu", menu);
	}

	/**
	 * gets the name of this entity
	 *
	 * @return name
	 */
	@Override
	public String getName() {
		return getRPClass().getName();
	}
}
